README.md
# mi-cpptest

This library is a intended as a minuscule unit-test library for C++11
and later:

- it should fail if a test fails
- it tries to print a message explaining the failure,
  - using `void mi_cpptest_stringify(std::ostream&, const T&)`, if defined,
  - else using `std::ostream& operator<<(std::ostream&, const T&)`, if defined,
  - else using a standard text
  to format values
- it tries to print TAP output
- it allows selecting which tests to run

The library has no dependencies beyond the standard C++ library.

## Running

At present, all command line arguments are interpreted test filters:

- `-{regex}` deselects tests matching `{regex}`
- `{regex}` selects tests matching `{regex}`
- searching the argument list stops at the first match, and the test
  is skipped if this was a regex with `-` prefix.

## Use with CMake

1. either include this as a subproject with `ADD_SUBDIRECTORY(...)`
2. or build and install and use `FIND_PACKAGE(mi-cpptest)`

In both cases, link to `mi-cpptest` or `mi-cpptest-main`.

## Use without CMake

As the library consists of a few files, it should be easy to use with
other build systems.


debian/changelog
mi-cpptest (0.2.0-1) unstable; urgency=low

  * new upstream version

 -- Alexander Bürger <alexander.buerger@met.no>  Fri, 21 May 2021 08:00:08 +0200

mi-cpptest (0.1.1-1) unstable; urgency=low

  * new upstream version

 -- Alexander Bürger <alexander.buerger@met.no>  Tue, 09 Mar 2021 08:00:08 +0100

mi-cpptest (0.1.0-1) unstable; urgency=low

  * debian packaging

 -- Alexander Bürger <alexander.buerger@met.no>  Wed, 24 Apr 2019 08:00:08 +0200


test/test_basic.cc
/*
  mi-cpptest

  Copyright (C) 2021 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of mi-cpptest.

  mi-cpptest is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  mi-cpptest is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mi-cpptest; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


#include "mi_cpptest.h"

#include <map>
#include <ostream>
#include <sstream>

#if 0
MI_CPPTEST_TEST_CASE(test_close)
{
    try {
        const double a=1, b=4, tol=1;
        MI_CPPTEST_REQUIRE_CLOSE(a, b, tol);
    } catch (miutil::cpptest::test_failure&) {
        return; // ok
    }
    MI_CPPTEST_FAIL("test must fail");
}

MI_CPPTEST_TEST_CASE(test_vector)
{
    const std::vector<std::string> a{"1"}, b{"1", "2"};
    MI_CPPTEST_CHECK_EQ(a, b);
}
#endif

MI_CPPTEST_TEST_CASE(test_relations) {
  MI_CPPTEST_REQUIRE_EQ(1, 1);
  MI_CPPTEST_REQUIRE_NE(1, 2);

  MI_CPPTEST_REQUIRE_GE(1, 1);
  MI_CPPTEST_REQUIRE_GT(2, 1);
  MI_CPPTEST_REQUIRE_LE(1, 2);
}

MI_CPPTEST_TEST_CASE(test_map) {
  const std::map<int, int> mapp{{1, 2}, {3, 4}};
  std::ostringstream out;
  out << miutil::cpptest::stringify(mapp);
  MI_CPPTEST_REQUIRE_EQ(out.str(), "{<1,2>,<3,4>}");
}

namespace {
struct NoOstream {
  int n;
};
} // namespace

MI_CPPTEST_TEST_CASE(test_type_without_ostream) {
  std::ostringstream oact;
  oact << miutil::cpptest::stringify(NoOstream{});
  const std::string act = oact.str();

  std::ostringstream exp;
  miutil::cpptest::mi_cpptest_stringify_missing(exp);

  MI_CPPTEST_REQUIRE_EQ(act.c_str(), exp.str());
}

namespace {
struct Stringify {
  int n;
};
void mi_cpptest_stringify(std::ostream &out, const Stringify &) {
  out << "stringify";
}
} // namespace

MI_CPPTEST_TEST_CASE(test_type_with_stringify) {
  std::ostringstream oact;
  oact << miutil::cpptest::stringify(Stringify{7});
  const std::string act = oact.str();

  MI_CPPTEST_REQUIRE_EQ(act, "stringify");
}

namespace {
struct OstreamAndStringify {
  int n;
};
std::ostream &operator<<(std::ostream &out, const OstreamAndStringify &) {
  return out << "ostream";
}
void mi_cpptest_stringify(std::ostream &out, const OstreamAndStringify &) {
  out << "stringify";
}
} // namespace

MI_CPPTEST_TEST_CASE(test_type_with_ostream_and_stringify) {
  std::ostringstream oact;
  oact << miutil::cpptest::stringify(OstreamAndStringify{7});
  const std::string act = oact.str();

  MI_CPPTEST_REQUIRE_EQ(act, "stringify");
}


test/test_with_fixture.cc
/*
  mi-cpptest

  Copyright (C) 2021 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of mi-cpptest.

  mi-cpptest is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  mi-cpptest is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mi-cpptest; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


#include "mi_cpptest.h"

namespace {
int up_ = 0, down_ = 0;

struct fixture : miutil::cpptest::test_fixture {
  void set_up() override { up_ += 1; }
  void tear_down() override { down_ += 1; }
};
} // namespace

MI_CPPTEST_FIXTURE_TEST_CASE(test_fixture, fixture) {
  MI_CPPTEST_CHECK_EQ(1, up_);
  MI_CPPTEST_CHECK_EQ(0, down_);
}

MI_CPPTEST_TEST_CASE(test_after_fixture) {
  MI_CPPTEST_CHECK_EQ(1, up_);
  MI_CPPTEST_CHECK_EQ(1, down_);
}


mi_cpptest.cc
/*
  mi-cpptest

  Copyright (C) 2019-2021 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of mi-cpptest.

  mi-cpptest is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  mi-cpptest is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mi-cpptest; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "mi_cpptest.h"

#include <iostream>
#include <regex>
#include <stdexcept>
#include <string>
#include <sstream>

namespace {

struct registered_test {
    std::string name;
    miutil::cpptest::test_function_t test;
    void operator()(miutil::cpptest::test_recorder *tr) const { test(tr); }
};

struct test_filter {
    bool exclusive;
    std::regex regex;
};

char hexchar(unsigned int i)
{
    const char hexchars[17] = "0123456789ABCDEF";
    return hexchars[i & 0xF];
}

void yaml_escaped_block(std::ostream &out, const std::string &text,
                        const std::string &indent) {
  out << indent;
  for (char ch : text) {
    if (ch == '\n') {
      out << '\n' << indent;
    } else if (ch < ' ' && ch != 9) {
      unsigned int u = (unsigned int)ch;
      out << "\\x" << hexchar(u >> 4) << hexchar(u);
    } else if (ch == '"' || ch == '\\') {
      out << '\\' << ch;
    } else {
      out << ch;
    }
  }
  out << '\n';
}

std::string strip_common_prefix(const std::string &from,
                                const std::string &prefix,
                                const std::string &strip) {
  size_t common = 0;
  const size_t max = std::min(from.size(), prefix.size());
  while (common < max && from[common] == prefix[common])
    ++common;
  const auto idx = from.find_first_not_of(strip, common);
  if (idx != std::string::npos)
    common = idx;
  return from.substr(common);
}

} // namespace

namespace miutil {
namespace cpptest {

void mi_cpptest_stringify_missing(std::ostream &out) {
  out << "???";
}

std::string test_recorder::file_prefix_;

void test_recorder::record(const char *file, int line,
                           const std::string &note) {
  std::ostringstream msg;
  if (line >= 0) {
    msg << strip_common_prefix(file, file_prefix_, "/:.") << ":" << line;
  }
  if (!note.empty()) {
    if (line >= 0)
      msg << ' ';
    msg << note;
  }
  messages_.push_back(msg.str());
}

test_status test_recorder::status() const {
  return messages_.empty() ? OK : FAIL;
}

typedef std::vector<registered_test> registered_test_v;

registered_test_v& registered_tests()
{
    static registered_test_v tests;
    return tests;
}

bool register_test(const char* name, test_function_t tf)
{
    registered_tests().push_back(registered_test {name, tf});
    return true;
}

void write_test_status(std::ostream &out, test_status status, size_t number,
                       const registered_test &rt, const std::string &message) {
  if (status == FAIL)
    out << "not ";
  out << "ok " << number;
  out << ' ' << rt.name;
  if (status == SKIP)
    out << " # SKIP";
  out << std::endl;
  if (!message.empty()) {
    out << " ---" << std::endl << " message: |" << std::endl;
    yaml_escaped_block(out, message, "   ");
    out << " ..." << std::endl;
  }
}

bool run_tests(size_t npatterns, char* patterns[])
{
    std::vector<test_filter> filters;
    filters.reserve(npatterns);
    size_t n_exclusive = 0;
    for (size_t i=0; i<npatterns; ++i) {
        char* pattern = patterns[i];
        test_filter tf;
        tf.exclusive = (pattern[0] == '-');
        if (tf.exclusive) {
            n_exclusive += 1;
            pattern += 1;
        }
        tf.regex = std::regex(pattern);
        filters.push_back(tf);
    }
    const bool use_default = (filters.empty() || n_exclusive == filters.size());

    std::cout << "TAP version 13" << std::endl
              << "1.." << registered_tests().size() << std::endl;
    bool all_passed = true;
    size_t test_number = 0;
    for (const registered_test& rt : registered_tests()) {
        test_number += 1;

        test_status status = use_default ? FAIL : SKIP;
        std::string message;

        std::smatch filter_match;
        for (const test_filter& tf : filters) {
            if (std::regex_match(rt.name, filter_match, tf.regex)) {
                status = tf.exclusive ? SKIP : FAIL;
                break;
            }
        }

        if (status != SKIP) {
          test_recorder tr;
          try {
            rt(&tr);
          } catch (const test_failure &tf) {
            // recorded in test_recorder::fail
            // tr.record(tf.file(), tf.line(), tf.message());
          } catch (std::exception &e) {
            tr.record("", -1, "uncaught exception: " + std::string(e.what()));
          } catch (...) {
            tr.record("", -1, "uncaught exception");
          }
          status = tr.status();
          if (status != OK) {
            all_passed = false;

            std::ostringstream msg;
            bool first = true;
            for (const auto &m : tr.messages()) {
              if (!first)
                msg << '\n';
              msg << m;
              first = false;
            }
            message = msg.str();
          }
        }
        write_test_status(std::cout, status, test_number, rt, message);
    }
    return all_passed;
}

bool run_tests_with_prefix(int argc, char *args[]) {
  test_recorder::set_file_prefix(args[0]);
  return run_tests(argc - 1, args + 1);
}

} // namespace cpptest
} // namespace miutil


mi_cpptest_main.cc
/*
  mi-cpptest

  Copyright (C) 2019-2021 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of mi-cpptest.

  mi-cpptest is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  mi-cpptest is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mi-cpptest; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "mi_cpptest.h"

int main(int argc, char* args[])
{
    return miutil::cpptest::run_tests(argc-1, args+1) ? 0 : 1;
}


mi_cpptest_version.h
/*
  mi-cpptest

  Copyright (C) 2019-2021 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of mi-cpptest.

  mi-cpptest is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  mi-cpptest is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mi-cpptest; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef MI_CPPTEST_VERSION_H
#define MI_CPPTEST_VERSION_H

#define MI_CPPTEST_VERSION_MAJOR 0
#define MI_CPPTEST_VERSION_MINOR 2
#define MI_CPPTEST_VERSION_PATCH 0

#define MI_CPPTEST_VERSION_INT(major,minor,patch) \
    (1000000*major + 1000*minor + patch)
#define MI_CPPTEST_VERSION_CURRENT_INT \
    MI_CPPTEST_VERSION_INT(MI_CPPTEST_VERSION_MAJOR, MI_CPPTEST_VERSION_MINOR, MI_CPPTEST_VERSION_PATCH)

#endif // MI_CPPTEST_VERSION_H


mi_cpptest.h
/* -*- c++ -*-
  mi-cpptest

  Copyright (C) 2019-2021 met.no

  Contact information:
  Norwegian Meteorological Institute
  Box 43 Blindern
  0313 OSLO
  NORWAY
  email: diana@met.no

  This file is part of mi-cpptest.

  mi-cpptest is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  mi-cpptest is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with mi-cpptest; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


#ifndef MI_CPPTEST_H
#define MI_CPPTEST_H

#include <cmath>
#include <cstdlib>
#include <functional>
#include <memory>
#include <ostream>
#include <sstream>
#include <type_traits>
#include <vector>

namespace miutil {
namespace cpptest {

// from https://en.cppreference.com/w/cpp/types/void_t
template <typename T, typename = void>
struct is_iterable : std::false_type {};
template <typename T>
struct is_iterable<T, std::void_t<decltype(std::declval<T>().begin()),
                                  decltype(std::declval<T>().end())>>
    : std::true_type {};

// from https://codereview.stackexchange.com/questions/250619/using-sfinae-to-provide-default-output-operators
template <typename T, typename = void>
struct has_output_operator : std::false_type {};
template <typename T>
struct has_output_operator<T, std::void_t<decltype(std::declval<std::ostream&>() << std::declval<T>())>> : std::true_type {};

void mi_cpptest_stringify_missing(std::ostream &out);

template <typename T, typename = void>
struct has_mi_cpptest_stringify : std::false_type {};
template <typename T>
struct has_mi_cpptest_stringify<T, std::void_t<decltype(mi_cpptest_stringify(std::declval<std::ostream&>(), std::declval<T>()))>> : std::true_type {};

template <class C, typename = void>
struct stringifier {
  stringifier(const C& c) : c_(c) {}
  void write(std::ostream &out) const { mi_cpptest_stringify_missing(out); }
  const C &c_;
};

template <class C>
stringifier<C> stringify(const C &c)
{
  return stringifier<C>(c);
}

template <class C>
std::ostream &operator<<(std::ostream &out, const stringifier<C> &s) {
  s.write(out);
  return out;
}

template <class C>
struct stringifier<C, typename std::enable_if<has_mi_cpptest_stringify<C>::value>::type> {
  stringifier(const C &c) : c_(c) {}
  void write(std::ostream &out) const { mi_cpptest_stringify(out, c_); }
  const C &c_;
};

template <class C>
struct stringifier<C, typename std::enable_if<has_output_operator<C>::value && !has_mi_cpptest_stringify<C>::value>::type> {
  stringifier(const C &c) : c_(c) {}
  void write(std::ostream &out) const { out << c_; }
  const C &c_;
};

template <class C>
struct stringifier<C, typename std::enable_if<is_iterable<C>::value && !has_output_operator<C>::value && !has_mi_cpptest_stringify<C>::value>::type> {
  stringifier(const C &c) : c_(c) {}
  void write(std::ostream &out) const {
    char sep = '{';
    for (const auto &e : c_) {
      out << sep << stringify(e);
      sep = ',';
    }
    out << '}';
  }
  const C &c_;
};

template <class A, class B>
struct stringifier<std::pair<A, B>> {
  typedef typename std::pair<A, B> P;
  stringifier(const P &p) : p_(p) {}
  void write(std::ostream &out) const {
    out << '<' << stringify(p_.first) << ',' << stringify(p_.second) << '>';
  }
  const P &p_;
};

struct test_failure : public std::exception {};

enum test_status { OK = 0, FAIL, SKIP };

class test_recorder {
public:
  void record(const char *file, int line,
              const std::string &msg = std::string());
  test_status status() const;
  const std::vector<std::string> &messages() const { return messages_; }

  static void set_file_prefix(const std::string &prefix) {
    file_prefix_ = prefix;
  }

private:
  static std::string file_prefix_;
  std::vector<std::string> messages_;
};

class test_fixture {
public:
  virtual void set_up() {}
  virtual void tear_down() {}

  test_recorder *mi_cpptest_recorder;
};

class test_fixture_up_down {
public:
  test_fixture_up_down(test_fixture &f) : fixture_(f) { fixture_.set_up(); }
  ~test_fixture_up_down() { fixture_.tear_down(); }

private:
  test_fixture &fixture_;
};

typedef void (*test_function_t)(test_recorder *);

bool register_test(const char* name, test_function_t tf);

bool run_tests(size_t npatterns, char* patterns[]);
bool run_tests_with_prefix(int argc, char *args[]);

template <class C> struct is_close {
  bool operator()(const C &a, const C &b, const C &tol) const {
    if (a == b)
      return true;
    const C d = std::abs(a - b), aa = std::abs(a), bb = std::abs(b);
    return (d <= aa * tol) && (d <= bb * tol);
  }
};

template <class C> struct is_near {
  bool operator()(const C &a, const C &b, const C &diff) const {
    return std::abs(a - b) < diff;
  }
};

} // namespace cpptest
} // namespace miutil

#define MI_CPPTEST_TEST_CASE(x)                                                \
  static void x(miutil::cpptest::test_recorder *);                             \
  static bool test4fimex_registered_##x =                                      \
      miutil::cpptest::register_test(#x, x);                                   \
  static void x(miutil::cpptest::test_recorder                                 \
                    *mi_cpptest_recorder) // { test body } after macro

#define MI_CPPTEST_FIXTURE_TEST_CASE(x, fixture)                               \
  static void tag_##x() {}                                                     \
  template <void (*F)()> struct x : public fixture { void run(); };            \
  static void run_##x(miutil::cpptest::test_recorder *tr) {                    \
    x<tag_##x> tf;                                                             \
    tf.mi_cpptest_recorder = tr;                                               \
    miutil::cpptest::test_fixture_up_down tear_down(tf);                       \
    tf.run();                                                                  \
  }                                                                            \
  static bool test4fimex_registered_##x =                                      \
      miutil::cpptest::register_test(#x, run_##x);                             \
  template <void (*F)()> void x<F>::run() // { test body } after macro

#define MI_CPPTEST_TEST_SUITE(x) // nothing
#define MI_CPPTEST_TEST_SUITE_END() // nothing

#define MI_CPPTEST___RECORD(fatal, x, m)                                       \
  do {                                                                         \
    if (!static_cast<bool>(x)) {                                               \
      std::ostringstream msg;                                                  \
      msg << m;                                                                \
      mi_cpptest_recorder->record(__FILE__, __LINE__, msg.str());              \
      if (fatal)                                                               \
        throw miutil::cpptest::test_failure();                                 \
    }                                                                          \
  } while (false)

#define MI_CPPTEST___RECORD_BOOL(fatal, x)                                     \
  MI_CPPTEST___RECORD(fatal, x, "'" #x "' does not evaluate to true")

#define MI_CPPTEST_REQUIRE_MESSAGE(x, m) MI_CPPTEST___RECORD(true, x, m)
#define MI_CPPTEST_REQUIRE(x) MI_CPPTEST___RECORD_BOOL(true, x)

#define MI_CPPTEST_CHECK_MESSAGE(x, m) MI_CPPTEST___RECORD(false, x, m)
#define MI_CPPTEST_CHECK(x) MI_CPPTEST___RECORD_BOOL(false, x)

#define MI_CPPTEST_FAIL(m) MI_CPPTEST___RECORD(true, false, "")

#define MI_CPPTEST___RECORD_OP2(fatal, x, y, op)                               \
  do {                                                                         \
    const auto xx = (x);                                                       \
    const auto yy = (y);                                                       \
    const auto ok = xx op yy;                                      \
    MI_CPPTEST___RECORD(fatal, ok,                                   \
                        #op " failed for " #x "="                              \
                            << miutil::cpptest::stringify(xx)                  \
                            << " and " #y "="                                  \
                            << miutil::cpptest::stringify(yy));                \
  } while (0)
#define MI_CPPTEST_REQUIRE_EQ(x, y)                                            \
  MI_CPPTEST___RECORD_OP2(true, x, y, ==)
#define MI_CPPTEST_CHECK_EQ(x, y)                                              \
  MI_CPPTEST___RECORD_OP2(false, x, y, ==)
#define MI_CPPTEST_REQUIRE_NE(x, y)                                            \
  MI_CPPTEST___RECORD_OP2(true, x, y, !=)
#define MI_CPPTEST_CHECK_NE(x, y)                                              \
  MI_CPPTEST___RECORD_OP2(false, x, y, !=)
#define MI_CPPTEST_REQUIRE_GT(x, y)                                            \
  MI_CPPTEST___RECORD_OP2(true, x, y, >)
#define MI_CPPTEST_CHECK_GT(x, y)                                              \
  MI_CPPTEST___RECORD_OP2(false, x, y, >)
#define MI_CPPTEST_REQUIRE_GE(x, y)                                            \
  MI_CPPTEST___RECORD_OP2(true, x, y, >=)
#define MI_CPPTEST_CHECK_GE(x, y)                                              \
  MI_CPPTEST___RECORD_OP2(false, x, y, >=)
#define MI_CPPTEST_REQUIRE_LT(x, y)                                            \
  MI_CPPTEST___RECORD_OP2(true, x, y, <)
#define MI_CPPTEST_CHECK_LT(x, y)                                              \
  MI_CPPTEST___RECORD_OP2(false, x, y, <)
#define MI_CPPTEST_REQUIRE_LE(x, y)                                            \
  MI_CPPTEST___RECORD_OP2(true, x, y, <=)
#define MI_CPPTEST_CHECK_LE(x, y)                                              \
  MI_CPPTEST___RECORD_OP2(false, x, y, <=)

#define MI_CPPTEST___RECORD_OP3(fatal, x, y, z, op)                            \
  do {                                                                         \
    const auto xx = (x);                                                       \
    const auto yy = (y);                                                       \
    const auto zz = (z);                                                       \
    const auto oper = op<decltype(xx)>{};                                      \
    MI_CPPTEST___RECORD(                                                       \
        fatal, oper(xx, yy, zz),                                               \
        #op " failed for " #x "="                                              \
            << miutil::cpptest::stringify(xx) << " and " #y "="                \
            << miutil::cpptest::stringify(yy) << " and " #z "="                \
            << miutil::cpptest::stringify(zz));                                \
  } while (0)

#define MI_CPPTEST___RECORD_CLOSE(fatal, x, y, z)                              \
  do {                                                                         \
    const auto xx = (x);                                                       \
    const auto yy = (y);                                                       \
    const auto zz = (z);                                                       \
    MI_CPPTEST___RECORD(fatal, miutil::cpptest::is_close(xx, yy, zz),          \
                        #x "='" << miutil::cpptest::stringify(xx)              \
                                << "' is not within tolerance " #z "='"        \
                                << miutil::cpptest::stringify(zz)              \
                                << "' from " #y "='"                           \
                                << miutil::cpptest::stringify(yy) << "'");     \
  } while (0)
#define MI_CPPTEST_REQUIRE_CLOSE(x, y, z)                                      \
  MI_CPPTEST___RECORD_OP3(true, x, y, z, miutil::cpptest::is_close)
#define MI_CPPTEST_CHECK_CLOSE(x, y, z)                                        \
  MI_CPPTEST___RECORD_OP3(false, x, y, z, miutil::cpptest::is_close)

#define MI_CPPTEST_REQUIRE_NEAR(x, y, z)                                       \
  MI_CPPTEST___RECORD_OP3(true, x, y, z, miutil::cpptest::is_near)
#define MI_CPPTEST_CHECK_NEAR(x, y, z)                                         \
  MI_CPPTEST___RECORD_OP3(false, x, y, z, miutil::cpptest::is_near)

#define MI_CPPTEST_CHECK_THROW(x, ex) \
    do { try { x; } catch (ex&) { break; } MI_CPPTEST_FAIL(); } while(0)
#define MI_CPPTEST_CHECK_NO_THROW(x) \
    do { try { x; } catch (...) { MI_CPPTEST_FAIL(); } } while(0)

#endif // MI_CPPTEST_H
